usb_test.go

```diff
@@ -94,3 +96,35 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
 		}
 	}
 }
+
+func TestOpenDeviceWithFileDescriptor(t *testing.T) {
+	ctx := newContextWithImpl(newFakeLibusb())
+	defer ctx.Close()
+
+	// file descriptor is an index to the FakeDevices array
+	for ix, d := range []struct {
+		vid, pid ID
+		exists   bool
+	}{
+		{0x9999, 0x0001, true},
+		{0x8888, 0x0002, true},
+		{0x1111, 0x1111, true},
+		{0x7777, 0x0003, false},
+	} {
+		// The fake keeps the order of fakeDevices so that it can be indexed by a fake "fileDescriptor"
+		dev, err := ctx.OpenDeviceWithFileDescriptor(uintptr(ix))
+		if (dev != nil) != d.exists {
+			t.Errorf("OpenDeviceWithFileDescriptor(%d): device != nil is %v, want %v", ix, dev != nil, d.exists)
+		}
+		if err != nil && d.exists { // It's OK to get an error if device doesn't exist
```

Comment:
the behavior of OpenDeviceWithFileDescriptor creates a small divergence:
the OpenDeviceWithVIDPID returns nil, nil if the device is not found. Here the function will return an error if the device is not found. I think this is fine though - the Open...VIDPID should have returned an error for not found too.

But this change makes it necessary to change the structure of the test, you cannot use the same structure - I think you don't need to distinguish "exists" vs "not exists", instead you need to distinguish errors vs non-errors.   And I think it may make sense to add a separate test case for a "want error", to simplify the code.

Reply:
Well, since the fakeDevice can have an uninitialised sysDevPtr (e.g. because the system doesn't support that?) there still is a case where a device exists, but is 'not useful'/'not good'. Perhaps the word 'exists' should rather be 'good' or 'useful' ?
I will split it in two test cases, this will actually make it avoid the wording problem and indeed make the cases simpler.